### PR TITLE
Strip any '+RTS ...' flags from command line options, remove GHC 8.2.2 support

### PR DESCRIPTION
**Pull request by csasarak**
Addresses #102 

I had a version of this in `addCmdOpts` but decided this was more appropriate. I can move it back if it isn't.

### CODE REVIEW

**Comment by mpickering** [comment]
Can you add a test please?

**Comment by csasarak** [comment]
Yes

**Comment by csasarak** [comment]
With respect to testing, I make new projects for each of stack/cabal and test against those, or add but I could also add a trivial `+RTS -A1m -RTS` to the existing ones. Is that sufficient? 

I can also add some more extensive tests of the function itself with different inputs

**Comment by mpickering** [comment]
Adding the options to the existing tests would be fine, adding a new test would also be fine, it doesn't matter too much which you choose.

File: src/HIE/Bios/Cradle.hs [hunk 1]
```diff
@@ -345,6 +346,16 @@ cabalAction work_dir mc l _fp = do
...
+-- Strip out any ["+RTS", ..., "-RTS"] sequences in the command string list.
+removeRTS :: [String] -> [String]
+removeRTS  = snd . foldr processPath (False, [])
+  where processPath p (inRTS, paths) =
+          case (inRTS, p) of
+            (_, "+RTS") -> (False, paths)
+            (_, "-RTS") -> (True, paths)
+            (True, _)   -> (inRTS, paths)
+            (_, p')     -> (False, p' : paths)
+
```
Comment (jinwoo) [inline src/HIE/Bios/Cradle.hs]: Sorry for a drive-by comment. Would code like this be simpler?

```haskell
removeRTS :: [String] -> [String]
removeRTS [] = []
removeRTS ("+RTS" : xs) =
  case dropWhile (/= "-RTS") xs of
    [] -> []
    (_ : ys) -> removeRTS ys
```

**Comment by csasarak** [comment]
I'm not at my computer at the moment, but as written i don't think this would work. The argument sequences also have other arguments in them besides the '+RTS...' bits which aren't necessarily at the beginning or end of the list.

If i'm understanding the proposal, I *think* your code will work with the addition of this case after your code.

```
removeRTS (y:ys) = y : removeRTS ys
```

I like the readability of what you wrote a lot and can change/test it again later. Thank you for having a look!


On December 19, 2019 4:58:37 PM EST, Jinwoo Lee <notifications@github.com> wrote:
>jinwoo commented on this pull request.
>
>
>
>> +-- Strip out any ["+RTS", ..., "-RTS"] sequences in the command
>string list.
>+removeRTS :: [String] -> [String]
>+removeRTS  = snd . foldr processPath (False, [])
>+  where processPath p (inRTS, paths) =
>+          case (inRTS, p) of
>+            (_, "+RTS") -> (False, paths)
>+            (_, "-RTS") -> (True, paths)
>+            (True, _)   -> (inRTS, paths)
>+            (_, p')     -> (False, p' : paths)
>+
>
>Sorry for a drive-by comment. Would code like this be simpler?
>
>```haskell
>removeRTS :: [String] -> [String]
>removeRTS [] = []
>removeRTS ("+RTS" : xs) =
>  case dropWhile (/= "-RTS") xs of
>    [] -> []
>    (_ : ys) -> removeRTS ys
>```

-- 
Chris

File: src/HIE/Bios/Cradle.hs (hunk 1 above)
Comment (csasarak) [reply src/HIE/Bios/Cradle.hs]: If i'm understanding the proposal, I *think* your code will work with the addition of this case after your code.

```
removeRTS (y:ys) = y : removeRTS ys
```

I like the readability of what you wrote a lot and can change/test it again later. Thank you for having a look!

Comment (jinwoo) [reply src/HIE/Bios/Cradle.hs]: Oops. Yeah, I forgot to add that case :)

Comment (csasarak) [reply src/HIE/Bios/Cradle.hs]: A new version is up.

**Comment by jinwoo** [comment]
Another drive-by comment :)

Per https://downloads.haskell.org/ghc/latest/docs/html/users_guide/runtime_control.html#rts-opts-cmdline, "No -RTS option is required if the runtime-system options extend to the end of the command line". It'd be good to add a test case like that too.

**Comment by csasarak** [comment]
@jinwoo I looked at this situation and cabal doesn't seem to build anything without the `-RTS`. When I checked, it's because additional non-rts ghc options were being added after `+RTS`. In my test it was adding `-hide-all-packages` which according to the [manual](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/packages.html#ghc-flag--hide-all-packages) is always passed by `cabal`. 

Even if GHC can accept the `+RTS ...` form on the command line, I don't think it makes sense in the context of a cabal file. I don't know of any way to control the order of options that cabal uses when invoking GHC, which would be necessary for the `+RTS ...` form to work reliably.

I couldn't think of what the test case for that would achieve then, since the cabal file seems to be unusable if you leave off the `-RTS` option. A cabal file where `hie-bios` works but cabal doesn't is maybe not an important case to support. But I want to hear if you agree with that, maybe there's a good reason for the additional test that I'm not thinking of. Or I could ask the cabal devs if that situation is valid or there's a way to mitigate it.

**Comment by jinwoo** [comment]
Oh I see. I had to do something similar in our environment and because we don't use cabal, I didn't realize it doesn't happen with cabal. Sorry about that.

I also realize that this change affects only cabal cradles. Why not make this change for all types of cradles? I guess it would benefit all of them.

Having said that, as I'm not the owner of this project, I'll just shut up now :)

Thank you for making this change!

**Comment by csasarak** [comment]
@jinwoo I see what you're saying and realize that I ignored the `bios` cradle type when writing tests. To make sure I've gotten everything:

* Stack: If you use stack then there's still a cabal file, and as far as I can tell it a stack cradle will still run things through `cabalWrapperArgs`. So stack cradles are supported by this code. Like regular cabal projects, stack won't build a project with `+RTS ...` form `ghc-options` in the cabal file.  There is a test as part of this PR for stack too.

* Cabal: This is supported, and exhibits the behavior where a project won't build if `ghc-options` has a `+RTS ...` form argument. Since cabal doesn't support it, I didn't worry about this case. 

* BIOS: This one I didn't have a test for, so thank you for prompting me to check. I don't see that anything was broken if you add +RTS options to the the script for this one. 

Can you let me know if there's a case here that isn't supported?

**Comment by csasarak** [comment]
So the current status of this is that the appveyor tests failed because the test RTS opts were being passed into the wrapper, which was exiting with an error because the RTS options GHC defaults to are 'some'. I changed the wrapper compilation code to include `-rtsopts=ignore` which just passes RTS options unchanged into the wrapper. It works for GHCs > 8.2.2, 8.2.2 and below don't seem to have an 'ignore' option. 

Setting `rtsopts=all` is wrong because it means the wrapper is eating the RTS options and not passing them to GHC underneath.

**Comment by mpickering** [comment]
We can probably just remove 8.2.2 support if it's causing issues. I think h-i-e or ghcide support that version.

**Comment by csasarak** [comment]
I think the simplest fix would be dropping support for 8.2.2 and using `-rtsopts=ignore`. I could turn on `rtsopts=all` for the `cabal.hs` program, but that would mean on windows the wrapper program catches those RTS opts and not the underlying call to `HIE_BIOS_GHC` which is what happens on linux. 

Can you clarify if this is correct: cabal or the `cabal.hs` program is receiving the same ghc-options as whatever component it's building. I'm guessing that that's because maybe a user might want some warnings for some components and not others, and the checks can be made to match what cabal would do at build-time for a given component. This is where the appveyor tests are failing I think, because the windows wrapper program doesn't accept the full complement of RTS options while cabal and stack do. 

Do you think that adding `-rtsopts=ignore` sufficient, or is that a bandaid?

**Comment by mpickering** [comment]
I think dropping support for 8.2.2 and using `-rtsopts=ignore` is a good solution. 

I don't understand your question sorry, `cabal` uses the compiled `cabal.hs` executable instead of `ghc` when invoked by `hie-bios`, this allows us to intercept the options it supplies to GHC to start a session.

**Comment by csasarak** [comment]
> I think dropping support for 8.2.2 and using `-rtsopts=ignore` is a good solution.

Sounds good, I will do that

> I don't understand your question sorry, `cabal` uses the compiled `cabal.hs` executable instead of `ghc` when invoked by `hie-bios`, this allows us to intercept the options it supplies to GHC to start a session.

I think I was just chasing my tail. I have a better understanding of how things work thanks to your comment.

**Comment by csasarak** [comment]
I think this is finished, but appveyor had a server error. I will try triggering a build again later.

**Comment by mpickering** [comment]
Needs a rebase and squash.

**Comment by csasarak** [comment]
@mpickering Let me know if there's anything I can do to further assist here. Thanks for your work on this project!
